File: src/test/java/de/zalando/zally/rules/PluralizeResourceNamesRuleTest.java
package de.zalando.zally.rules;

import io.swagger.models.Swagger;
import io.swagger.parser.SwaggerParser;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

public class PluralizeResourceNamesRuleTest {
    private Swagger validSwagger = getFixture("fixtures/pluralizeResourcesValid.json");
    private Swagger invalidSwagger = getFixture("fixtures/pluralizeResourcesInvalid.json");


    @Test
    public void positiveCase() {
        assertThat(new PluralizeResourceNamesRule().validate(validSwagger)).isEmpty();
    }

    @Test
    public void negativeCase () {
        List<String> results = new PluralizeResourceNamesRule()
                .validate(invalidSwagger)
                .stream()
                .map(v -> v.getPath().get())
                .collect(Collectors.toList());
        assertThat(results).hasSameElementsAs(Arrays.asList("/pet/cats", "/pets/cats/{cat-id}/tail/{tail-id}/strands"));
    }

    @Test
    public void positiveCaseSpp() {
        Swagger swagger = getFixture("api_spp.json");
        assertThat(new PluralizeResourceNamesRule().validate(swagger)).isEmpty();
    }

    @Test
    public void positiveCaseTinbox() {
        Swagger swagger = getFixture("api_tinbox.yaml");
        assertThat(new PluralizeResourceNamesRule().validate(swagger)).isEmpty();
    }

    private Swagger getFixture(String fixture) {
        return new SwaggerParser().read(fixture);
    }
}

File: server/src/test/java/de/zalando/zally/rule/PluralizeResourceNamesRuleTest.kt
package de.zalando.zally.rule

import de.zalando.zally.getFixture
import de.zalando.zally.testConfig
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class PluralizeResourceNamesRuleTest {

    @Test
    fun positiveCase() {
        val swagger = getFixture("pluralizeResourcesValid.json")
        assertThat(PluralizeResourceNamesRule(testConfig).validate(swagger)).isNull()
    }

    @Test
    fun negativeCase() {
        val swagger = getFixture("pluralizeResourcesInvalid.json")
        val result = PluralizeResourceNamesRule(testConfig).validate(swagger)!!
        assertThat(result.paths).hasSameElementsAs(listOf("/pet/cats", "/pets/cats/{cat-id}/tail/{tail-id}/strands"))
    }

    @Test
    fun positiveCaseSpp() {
        val swagger = getFixture("api_spp.json")
        assertThat(PluralizeResourceNamesRule(testConfig).validate(swagger)).isNull()
    }

    @Test
    fun positiveCasePathsWithTheApiPrefix() {
        val swagger = getFixture("spp_with_paths_having_api_prefix.json")
        assertThat(PluralizeResourceNamesRule(testConfig).validate(swagger)).isNull()
    }

    @Test
    fun negativeCaseTinbox() {
        val swagger = getFixture("api_tinbox.yaml")
        val result = PluralizeResourceNamesRule(testConfig).validate(swagger)!!
        assertThat(result.paths).hasSameElementsAs(listOf("/queue/configs/{config-id}", "/queue/models",
            "/queue/models/{model-id}", "/queue/summaries"))
    }

}
